### PR TITLE
Feature/flag quiet

### PR DESCRIPTION
**Pull request by ashupatil96**
## Description

Sometimes developers choose to not see all logs/output of your command and just see success or failure as the outcome. Hence adding --quiet or -q flag to 1build command execution will result in just executing the command but not showing the entire output of it.

## New dependencies introduced
-   [Viper](https://github.com/spf13/viper) 
     Reason: Viper has the ability to bind to flags. To bind flag --quiet flag throughout project(particularly  exec.go)

## Checklist

-   [N] I have added tests that prove my fix is effective or that my feature works
-   [Y] I have update README (If needed)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,6 +140,17 @@ you can use one of them, both or neither.
...
 
+### Using `--quiet` or `-q` flag
+Sometimes you choose to not see all logs/output of your command and just see success or failure as the outcome. 
+So using `--quite` or `-q` flag to 1build command execution will result in just executing the command but not showing the entire output of it, Only shows SUCCESS/FAILURE as result of command execution.
+```console
+  1build lint test --quiet
+  ```
+  OR
+```console
+  1build lint test --q
```
Comment (gopinath-langote) [inline README.md]:
```suggestion
  1build lint test -q
```

File: cmd/exec/exec.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	"github.com/gopinath-langote/1build/cmd/config"
 	"github.com/gopinath-langote/1build/cmd/models"
 	"github.com/gopinath-langote/1build/cmd/utils"
+	"github.com/spf13/viper"
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: Any reason to introduce this dependency? 

As we are using cobra default flags which is working fine for us.

File: cmd/exec/exec.go [hunk 3]
```diff
@@ -42,15 +43,24 @@ func ExecutePlan(commands ...string) {
 }
 
 func executeAndStopIfFailed(command *models.CommandContext, executeStart time.Time) {
-	command.PrintPhaseBanner()
-	err := command.CommandSession.Run()
-	if err != nil {
+	stdout, err := command.CommandSession.CombinedOutput()
+	if !viper.GetBool("quiet") {
+		command.PrintPhaseBanner()
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: Let's print phase banner regardless of quiet flag - As you would be interested in seeing which phase failed amount all

File: cmd/exec/exec.go [hunk 4]
```diff
@@ -42,15 +43,24 @@ func ExecutePlan(commands ...string) {
 }
 
 func executeAndStopIfFailed(command *models.CommandContext, executeStart time.Time) {
-	command.PrintPhaseBanner()
-	err := command.CommandSession.Run()
-	if err != nil {
+	stdout, err := command.CommandSession.CombinedOutput()
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: With this - will you get realtime output when you dont specify flag.

Example - if you run gradle task which takes 2 min to finish - Will this show realtime output logs of gradle or entire log will be visible after 2 min?

We want the first one - see realtime output

File: cmd/exec/exec.go [hunk 5]
```diff
@@ -42,15 +43,24 @@ func ExecutePlan(commands ...string) {
...
+		command.PrintPhaseBanner()
+		if err != nil {
+			exitCode := (err.Error())[12:]
+			text := "\nExecution failed in phase '" + command.Name + "' – exit code: " + exitCode
+			utils.CPrintln(text, utils.Style{Color: utils.RED})
+			printResultsBanner(false, executeStart)
+			utils.ExitWithCode(exitCode)
+		} else {
+			fmt.Printf("%s", stdout)
+		}
```
Comment (gopinath-langote) [inline cmd/exec/exec.go]: Test cases are missing for the same. Could you please write test? Let me know if you need any help

File: cmd/root.go [hunk 6]
```diff
@@ -39,4 +41,7 @@ func Execute() {
 
 func init() {
 	rootCmd.SetHelpCommand(&cobra.Command{Use: "no-help", Hidden: true})
+	rootCmd.PersistentFlags().BoolP("quiet", "q", false,
+		"Only shows SUCCESS/FAILURE as result of command execution instead of showing all output of command")
```
Comment (gopinath-langote) [inline cmd/root.go]: Description could be better - 

What do you think about `Hide output log of command & only show SUCCESS/FAILURe result`?

File: README.md [hunk 7]
```diff
@@ -140,6 +140,17 @@ you can use one of them, both or neither.
     1build unset before after
     ```
 
+### Using `--quiet` or `-q` flag
+Sometimes you choose to not see all logs/output of your command and just see success or failure as the outcome. 
+So using `--quite` or `-q` flag to 1build command execution will result in just executing the command but not showing the entire output of it, Only shows SUCCESS/FAILURE as result of command execution.
```
Comment (gopinath-langote) [inline README.md]: Codacy Issue found: [[maximum-line-length] Line must be at most 190 characters](https://app.codacy.com/manual/gopinath-langote/1build/pullRequest?prid=5214834)

**Comment by gopinath-langote** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gopinath-langote/1build/pullRequest?prid=5214834&bid=17099351)

File: cmd/exec/exec.go (hunk 2 above)
Comment (ashupatil96) [reply cmd/exec/exec.go]: I faced 2 issues  while accessing --quiet flag in exec.go(exec package). So this dependency added.
1) as I added PersistentFlags() to rootCmd (cmd/root.go) , and rootCmd is private identifier,
So rootCmd is not available outside `cmd` pacakge (rootCmd is required to get flag value),
and we need to access flag(--quiet) in `exec` package.
2) Even if we make, rootCmd as global identifier , and when we access  rootCmd from exec.go, Problem of cyclic import arises.

File: cmd/exec/exec.go (hunk 4 above)
Comment (ashupatil96) [reply cmd/exec/exec.go]: will change implementation.

File: cmd/root.go (hunk 6 above)
Comment (ashupatil96) [reply cmd/root.go]: will change description.

File: cmd/exec/exec.go (hunk 5 above)
Comment (ashupatil96) [reply cmd/exec/exec.go]: i will add test cases.

File: cmd/exec/exec.go (hunk 2 above)
Comment (ashupatil96) [reply cmd/exec/exec.go]: @gopinath-langote what do you think of this ?

Comment (gopinath-langote) [reply cmd/exec/exec.go]: I see. Circular dependency we definitely don't want.

- Can we parse the parameter (flags) in root command pass those to exec?

File: README.md (hunk 7 above)
Comment (gopinath-langote) [reply README.md]: Please address this too

**Comment by gopinath-langote** [comment]
Let me know if you need any help from my side.

File: cmd/exec/exec.go (hunk 2 above)
Comment (ashupatil96) [reply cmd/exec/exec.go]: yes, we can pass parameter in rootCmd Run Function (root.go).
like .  _**exec.ExecutePlan(quietFlag, args...)**_
1) we need to update parameters in exec.go functions also.
     a) ExecutePlan().  -> from root.go this function is called
     b) executeAndStopIfFailed(). -> from ExecutePlan() this function is called.
         and this function is called 3 times in exec.go
2) Even in future also, if we need to add more flags like this then parsing flags and passing to function, might become complex, as we need to pass every flag to function. 

@gopinath-langote what do you think ? should I use this implementation of parsing  parameters ?

Comment (gopinath-langote) [reply cmd/exec/exec.go]: @ashupatil96 Thanks for the detailed explanation. 

As the parsing implementation needs lot of effort and also makes code complex.

Let's go ahead with your implementation using `viper`

**Comment by ashupatil96** [comment]
@gopinath-langote Please Review.

**Comment by gopinath-langote** [comment]
@ashupatil96 LGTM.

Thanks for the PR. This was a long waiting feature for me.

**Pull request merged**